File: server/proxy.js
require('dotenv').config();

const http = require('http');
const WebSocket = require('ws');
const { URL } = require('url');
const fs = require('fs');
const path = require('path');

// ========== LOGGING SYSTEM ==========
const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    dim: '\x1b[2m',
    underscore: '\x1b[4m',
    blink: '\x1b[5m',
    reverse: '\x1b[7m',
    hidden: '\x1b[8m',
    
    fg: {
        black: '\x1b[30m',
        red: '\x1b[31m',
        green: '\x1b[32m',
        yellow: '\x1b[33m',
        blue: '\x1b[34m',
        magenta: '\x1b[35m',
        cyan: '\x1b[36m',
        white: '\x1b[37m',
        gray: '\x1b[90m',
        crimson: '\x1b[38m'
    },
    bg: {
        black: '\x1b[40m',
        red: '\x1b[41m',
        green: '\x1b[42m',
        yellow: '\x1b[43m',
        blue: '\x1b[44m',
        magenta: '\x1b[45m',
        cyan: '\x1b[46m',
        white: '\x1b[47m'
    }
};

const LOG_LEVELS = {
    ERROR: 0,
    WARN: 1,
    INFO: 2,
    DEBUG: 3
};

const LOG_CONFIG = {
    level: LOG_LEVELS[process.env.LOG_LEVEL?.toUpperCase() || 'INFO'] || LOG_LEVELS.INFO,
    enableColors: process.env.LOG_COLORS !== 'false',
    enableFile: process.env.LOG_FILE === 'true',
    logPath: process.env.LOG_PATH || path.join(process.cwd(), 'logs'),
    consoleCompact: process.env.LOG_CONSOLE_COMPACT !== 'false'
};

// Create log directory if file logging is enabled
if (LOG_CONFIG.enableFile && !fs.existsSync(LOG_CONFIG.logPath)) {
    fs.mkdirSync(LOG_CONFIG.logPath, { recursive: true });
}

function getTimestamp() {
    const now = new Date();
    return now.toISOString().replace('T', ' ').substring(0, 19);
}

function getShortTimestamp() {
    const now = new Date();
    return now.toLocaleTimeString('en-US', { hour12: false });
}

function colorize(text, color) {
    if (!LOG_CONFIG.enableColors) return text;
    return `${color}${text}${colors.reset}`;
}

function getLevelColor(level) {
    switch(level) {
        case 'ERROR': return colors.fg.red;
        case 'WARN': return colors.fg.yellow;
        case 'INFO': return colors.fg.green;
        case 'DEBUG': return colors.fg.cyan;
        default: return colors.fg.white;
    }
}

function getLevelBgColor(level) {
    switch(level) {
        case 'ERROR': return colors.bg.red;
        case 'WARN': return colors.bg.yellow;
        case 'INFO': return colors.bg.green;
        case 'DEBUG': return colors.bg.cyan;
        default: return colors.bg.white;
    }
}

function writeToFile(level, message, fullMessage) {
    if (!LOG_CONFIG.enableFile) return;
    
    const date = new Date();
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    const logFile = path.join(LOG_CONFIG.logPath, `proxy-${year}-${month}-${day}.log`);
    
    const logEntry = `[${getTimestamp()}] [${level}] ${fullMessage}\n`;
    fs.appendFileSync(logFile, logEntry, 'utf8');
}

function log(level, message, ...args) {
    const levelValue = LOG_LEVELS[level];
    if (levelValue > LOG_CONFIG.level) return;
    
    const timestamp = LOG_CONFIG.consoleCompact ? getShortTimestamp() : getTimestamp();
    const levelColor = getLevelColor(level);
    const coloredLevel = colorize(level.padEnd(5), levelColor);
    
    let formattedMessage = message;
    if (args.length > 0) {
        formattedMessage = message.replace(/{}/g, () => {
            const arg = args.shift();
            return typeof arg === 'object' ? JSON.stringify(arg) : arg;
        });
        if (args.length > 0) {
            formattedMessage += ' ' + args.map(a => typeof a === 'object' ? JSON.stringify(a) : a).join(' ');
        }
    }
    
    const consoleMessage = LOG_CONFIG.consoleCompact 
        ? `${coloredLevel} ${timestamp} ${formattedMessage}`
        : `${coloredLevel} [${timestamp}] ${formattedMessage}`;
    
    console.log(consoleMessage);
    writeToFile(level, formattedMessage, `[${timestamp}] ${formattedMessage}`);
}

// Create logger object
const logger = {
    error: (msg, ...args) => log('ERROR', msg, ...args),
    warn: (msg, ...args) => log('WARN', msg, ...args),
    info: (msg, ...args) => log('INFO', msg, ...args),
    debug: (msg, ...args) => log('DEBUG', msg, ...args),
    
    // Special formatted logs
    connection: (id, direction, status, details = '') => {
        const statusColor = status === 'connected' ? colors.fg.green : colors.fg.red;
        const statusSymbol = status === 'connected' ? '▲' : '▼';
        logger.info(`${colorize(statusSymbol, statusColor)} ${direction.padEnd(7)} ${colorize(id, colors.fg.cyan)} ${status}${details ? ' ' + details : ''}`);
    },
    
    proxy: (from, to, type, id, data = '') => {
        const typeColor = type === 'msg' ? colors.fg.magenta : colors.fg.blue;
        const arrow = colorize('→', colors.fg.gray);
        logger.debug(`${colorize(from.padEnd(6), colors.fg.gray)} ${arrow} ${colorize(to.padEnd(6), colors.fg.gray)} ${colorize(type, typeColor)} ${colorize(id, colors.fg.cyan)}${data ? ' ' + data.substring(0, 50) : ''}`);
    },
    
    divider: (char = '=', length = 50) => {
        if (!LOG_CONFIG.consoleCompact) {
            console.log(colorize(char.repeat(length), colors.fg.gray));
        }
    }
};

// ========== PROXY SERVER ==========
const LOCAL_PORT = process.env.LOCAL_PORT || 8080;
const REMOTE_URL = process.env.REMOTE_URL;

if (!REMOTE_URL) {
    logger.error('REMOTE_URL environment variable is required');
    process.exit(1);
}

// Store local clients and their mappings to remote connections
const localClients = new Map();
let nextLocalId = 1;

function createLocalId() {
    return 'L' + (nextLocalId++) + '_' + Date.now();
}

// Create HTTP server for status page and WebSocket upgrade
const server = http.createServer((req, res) => {
    if (req.method === 'GET' && req.url === '/') {
        res.writeHead(200, { 'Content-Type': 'text/html' });
        res.write(`<!DOCTYPE html>
        <html>
        <head>
            <title>TheCards Local Proxy</title>
            <style>
                body { font-family: monospace; padding: 2rem; background: #1e1e1e; color: #d4d4d4; }
                h1 { color: #4ec9b0; }
                .status { background: #2d2d2d; padding: 1rem; border-radius: 5px; margin: 1rem 0; }
                .connected { color: #4ec9b0; }
                .disconnected { color: #f48771; }
                .client { background: #252526; padding: 0.5rem; margin: 0.5rem 0; border-left: 3px solid #4ec9b0; }
                pre { margin: 0; }
            </style>
        </head>
        <body>
            <h1>TheCards Local Proxy</h1>
            <div class="status">
                <strong>Status:</strong> <span class="connected">RUNNING</span><br>
                <strong>Remote Server:</strong> ${REMOTE_URL}<br>
                <strong>Active Clients:</strong> <span id="clientCount">${localClients.size}</span>
            </div>
            <h2>Active Connections</h2>
            <div id="clients">
                ${Array.from(localClients.entries()).map(([id, data]) => `
                    <div class="client">
                        <pre>ID: ${id}<br>Room: ${data.roomId || 'None'}<br>Host: ${data.hostId === id ? 'Yes' : 'No'}</pre>
                    </div>
                `).join('') || '<p>No active connections</p>'}
            </div>
            <p style="margin-top: 2rem; font-size: 0.8rem;">Connect using: <strong>ws://localhost:8080</strong></p>
        </body>
        </html>`);
        res.end();
    } else if (req.method === 'GET' && req.url === '/status') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({
            status: 'running',
            remoteServer: REMOTE_URL,
            activeClients: localClients.size,
            clients: Array.from(localClients.entries()).map(([id, data]) => ({
                id,
                roomId: data.roomId,
                isHost: data.hostId === id
            }))
        }));
    } else {
        res.writeHead(404);
        res.end('Not found');
    }
});

// WebSocket server for local clients
const wss = new WebSocket.Server({ server });

wss.on('connection', (localSocket, req) => {
    const localClientId = createLocalId();
    logger.connection(localClientId, 'LOCAL', 'connected', `from ${req.socket.remoteAddress}`);
    
    const clientState = {
        localSocket,
        remoteSocket: null,
        roomId: null,
        hostId: null,
        messageQueue: []
    };
    
    localClients.set(localClientId, clientState);
    
    logger.debug(`Connecting to remote server for {}`, localClientId);
    
    let remoteSocket;
    try {
        remoteSocket = new WebSocket(REMOTE_URL);
        clientState.remoteSocket = remoteSocket;
    } catch (e) {
        logger.error(`Failed to create remote connection for {}: {}`, localClientId, e.message);
        localSocket.close(1011, "Internal Error");
        localClients.delete(localClientId);
        return;
    }
    
    remoteSocket.on('open', () => {
        logger.connection(localClientId, 'REMOTE', 'connected');
        
        while (clientState.messageQueue.length > 0) {
            const msgData = clientState.messageQueue.shift();
            try {
                remoteSocket.send(msgData);
                logger.proxy('QUEUE', 'REMOTE', 'flush', localClientId);
            } catch(e) {
                logger.error(`Error sending queued message: {}`, e.message);
            }
        }
    });
    
    remoteSocket.on('message', (data) => {
        try {
            const dataStr = data.toString();
            const msg = JSON.parse(dataStr);
            logger.proxy('REMOTE', 'LOCAL', msg.type || 'MSG', localClientId, dataStr.substring(0, 100));
            
            if (msg.type === 'ROOM_CREATED') {
                clientState.roomId = msg.code;
                clientState.hostId = localClientId;
                logger.info(`Room created: {} (host: {})`, msg.code, localClientId);
            } else if (msg.type === 'JOIN_SUCCESS') {
                clientState.roomId = msg.code;
                logger.info(`Joined room: {} (client: {})`, msg.code, localClientId);
            } else if (msg.type === 'ROOM_CLOSED') {
                clientState.roomId = null;
                clientState.hostId = null;
                logger.warn(`Room closed for client {}`, localClientId);
            }
            
            if (localSocket.readyState === WebSocket.OPEN) {
                localSocket.send(dataStr);
            }
        } catch (e) {
            logger.debug(`Forwarding raw message REMOTE → LOCAL`, localClientId);
            if (localSocket.readyState === WebSocket.OPEN) {
                localSocket.send(data.toString());
            }
        }
    });
    
    remoteSocket.on('close', (code, reason) => {
        logger.connection(localClientId, 'REMOTE', 'disconnected', `code:${code} reason:${reason || 'none'}`);
        
        if (localSocket.readyState === WebSocket.OPEN) {
            try {
                localSocket.send(JSON.stringify({
                    type: 'ERROR',
                    message: 'Connection to remote server lost'
                }));
                localSocket.close(1011, "Remote connection lost");
            } catch(e) {}
        }
        
        localClients.delete(localClientId);
    });
    
    remoteSocket.on('error', (err) => {
        logger.error(`Remote socket error for {}: {}`, localClientId, err.message);
    });
    
    localSocket.on('message', (data) => {
        try {
            const dataStr = data.toString();
            try {
                const msg = JSON.parse(dataStr);
                logger.proxy('LOCAL', 'REMOTE', msg.type || 'MSG', localClientId, dataStr.substring(0, 100));
            } catch (e) {
                logger.proxy('LOCAL', 'REMOTE', 'RAW', localClientId, dataStr.substring(0, 100));
            }
            
            if (remoteSocket.readyState === WebSocket.OPEN) {
                remoteSocket.send(dataStr);
            } else if (remoteSocket.readyState === WebSocket.CONNECTING) {
                logger.debug(`Queuing message for remote (state: CONNECTING)`, localClientId);
                clientState.messageQueue.push(dataStr);
            } else {
                logger.warn(`Dropping message - remote not connected (state: {})`, remoteSocket.readyState);
            }
        } catch (e) {
            logger.error(`Error handling local message: {}`, e.message);
        }
    });
    
    localSocket.on('close', (code, reason) => {
        logger.connection(localClientId, 'LOCAL', 'disconnected', `code:${code}`);
        if (remoteSocket && (remoteSocket.readyState === WebSocket.OPEN || remoteSocket.readyState === WebSocket.CONNECTING)) {
            remoteSocket.close(1000, "Local client disconnected");
        }
        localClients.delete(localClientId);
    });
    
    localSocket.on('error', (err) => {
        logger.error(`Local socket error for {}: {}`, localClientId, err.message);
    });
});

// Start the proxy server
server.listen(LOCAL_PORT, '0.0.0.0', () => {
    console.log('\n' + colorize('═'.repeat(60), colors.fg.cyan));
    console.log(colorize('  TheCards Local Proxy Server', colors.bright + colors.fg.green));
    console.log(colorize('═'.repeat(60), colors.fg.cyan));
    console.log(`  ${colorize('Local WS:', colors.fg.yellow)}  ws://0.0.0.0:${LOCAL_PORT}`);
    console.log(`  ${colorize('Local HTTP:', colors.fg.yellow)} http://0.0.0.0:${LOCAL_PORT}`);
    console.log(`  ${colorize('Remote:', colors.fg.yellow)}     ${REMOTE_URL}`);
    console.log(`  ${colorize('Log Level:', colors.fg.yellow)}  ${Object.keys(LOG_LEVELS).find(k => LOG_LEVELS[k] === LOG_CONFIG.level) || 'INFO'}`);
    console.log(`  ${colorize('File Logging:', colors.fg.yellow)} ${LOG_CONFIG.enableFile ? colorize('ENABLED', colors.fg.green) : colorize('DISABLED', colors.fg.gray)}`);
    if (LOG_CONFIG.enableFile) {
        console.log(`  ${colorize('Log Path:', colors.fg.yellow)}     ${LOG_CONFIG.logPath}`);
    }
    console.log(colorize('═'.repeat(60), colors.fg.cyan));
    console.log(`\n  ${colorize('OK', colors.fg.green)} Proxy running! Connect your game client to:`);
    console.log(`   ${colorize(`ws://localhost:${LOCAL_PORT}`, colors.bright + colors.fg.cyan)}`);
    console.log(`  ${colorize('i', colors.fg.blue)} Status page: ${colorize(`http://localhost:${LOCAL_PORT}`, colors.fg.cyan)}`);
    console.log(colorize('═'.repeat(60), colors.fg.cyan) + '\n');
});

// Handle process termination
process.on('SIGINT', () => {
    logger.info('\nShutting down proxy...');
    
    for (const [id, data] of localClients) {
        if (data.remoteSocket && data.remoteSocket.readyState === WebSocket.OPEN) {
            data.remoteSocket.close();
        }
        if (data.localSocket && data.localSocket.readyState === WebSocket.OPEN) {
            data.localSocket.close();
        }
    }
    
    server.close(() => {
        logger.info('Proxy stopped.');
        process.exit(0);
    });
});

File: server/proxy.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/gorilla/websocket"
)

// ClientState represents a connected local client
type ClientState struct {
	localSocket   *websocket.Conn
	remoteSocket  *websocket.Conn
	roomID        string
	hostID        string
	messageQueue  [][]byte
	mu            sync.Mutex
}

// Config holds application configuration
type Config struct {
	LocalPort string
	RemoteURL string
}

var (
	config       Config
	localClients sync.Map // string -> *ClientState
	nextLocalID  uint64
	upgrader     = websocket.Upgrader{
		CheckOrigin: func(r *http.Request) bool { return true },
	}
	exePath, _ = os.Executable()
	exeDir     = filepath.Dir(exePath)
	exeName    = strings.TrimSuffix(filepath.Base(exePath), filepath.Ext(exePath))
	configPath = filepath.Join(exeDir, exeName+".ini")
)

func createLocalID() string {
	id := atomic.AddUint64(&nextLocalID, 1)
	return fmt.Sprintf("L%d_%d", id, time.Now().UnixNano())
}

func loadConfig() error {
	config.LocalPort = "8080"
	config.RemoteURL = ""

	file, err := os.Open(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(line, ";") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			continue
		}
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])

		switch strings.ToUpper(key) {
		case "LOCAL_PORT":
			config.LocalPort = value
		case "REMOTE_URL":
			config.RemoteURL = value
		}
	}
	return scanner.Err()
}

func saveConfig() error {
	content := fmt.Sprintf(`# DevilBridge Local Proxy Configuration
LOCAL_PORT=%s
REMOTE_URL=%s
`, config.LocalPort, config.RemoteURL)
	return os.WriteFile(configPath, []byte(content), 0644)
}

func handleStatusPage(w http.ResponseWriter, r *http.Request) {
	if r.Method != "GET" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	if r.URL.Path == "/" {
		w.Header().Set("Content-Type", "text/html")
		
		clientsHTML := ""
		count := 0
		localClients.Range(func(key, value interface{}) bool {
			id := key.(string)
			state := value.(*ClientState)
			roomID := state.roomID
			if roomID == "" {
				roomID = "None"
			}
			isHost := "No"
			if state.hostID == id {
				isHost = "Yes"
			}
			clientsHTML += fmt.Sprintf(`
                    <div class="client">
                        <pre>ID: %s<br>Room: %s<br>Host: %s</pre>
                    </div>`, id, roomID, isHost)
			count++
			return true
		})

		if clientsHTML == "" {
			clientsHTML = "<p>No active connections</p>"
		}

		html := fmt.Sprintf(`<!DOCTYPE html>
        <html>
        <head>
            <title>DevilBridge Local Proxy</title>
            <style>
                body { font-family: monospace; padding: 2rem; background: #1e1e1e; color: #d4d4d4; }
                h1 { color: #4ec9b0; }
                .status { background: #2d2d2d; padding: 1rem; border-radius: 5px; margin: 1rem 0; }
                .connected { color: #4ec9b0; }
                .disconnected { color: #f48771; }
                .client { background: #252526; padding: 0.5rem; margin: 0.5rem 0; border-left: 3px solid #4ec9b0; }
                pre { margin: 0; }
            </style>
            <meta http-equiv="refresh" content="5">
        </head>
        <body>
            <h1>DevilBridge Local Proxy</h1>
            <div class="status">
                <strong>Status:</strong> <span class="connected">● RUNNING</span><br>
                <strong>Remote Server:</strong> %s<br>
                <strong>Active Clients:</strong> %d
            </div>
            <h2>Active Connections</h2>
            <div id="clients">
                %s
            </div>
            <p style="margin-top: 2rem; font-size: 0.8rem;">Connect using: <strong>ws://localhost:%s</strong></p>
        </body>
        </html>`, config.RemoteURL, count, clientsHTML, config.LocalPort)
		
		fmt.Fprint(w, html)
	} else if r.URL.Path == "/status" {
		w.Header().Set("Content-Type", "application/json")
		
		clients := []map[string]interface{}{}
		localClients.Range(func(key, value interface{}) bool {
			id := key.(string)
			state := value.(*ClientState)
			clients = append(clients, map[string]interface{}{
				"id":     id,
				"roomId": state.roomID,
				"isHost": state.hostID == id,
			})
			return true
		})

		response := map[string]interface{}{
			"status":        "running",
			"remoteServer":  config.RemoteURL,
			"activeClients": len(clients),
			"clients":       clients,
		}
		json.NewEncoder(w).Encode(response)
	} else {
		http.NotFound(w, r)
	}
}

func handleWebSocket(w http.ResponseWriter, r *http.Request) {
	localSocket, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("[ERROR] Failed to upgrade connection: %v", err)
		return
	}

	localClientID := createLocalID()
	log.Printf("[LOCAL] New client connected: %s", localClientID)

	state := &ClientState{
		localSocket:  localSocket,
		remoteSocket: nil,
		roomID:       "",
		hostID:       "",
		messageQueue: make([][]byte, 0),
	}
	localClients.Store(localClientID, state)

	log.Printf("[REMOTE] Connecting to remote server for %s...", localClientID)

	remoteSocket, _, err := websocket.DefaultDialer.Dial(config.RemoteURL, nil)
	if err != nil {
		log.Printf("[REMOTE] Failed to create connection for %s: %v", localClientID, err)
		localSocket.Close()
		localClients.Delete(localClientID)
		return
	}
	state.remoteSocket = remoteSocket

	// Handle remote -> local messages
	go func() {
		defer func() {
			remoteSocket.Close()
			localSocket.Close()
			localClients.Delete(localClientID)
		}()

		for {
			_, message, err := remoteSocket.ReadMessage()
			if err != nil {
				log.Printf("[REMOTE] Disconnected for client %s: %v", localClientID, err)
				
				// Send error to local client before closing
				errMsg := map[string]string{
					"type":    "ERROR",
					"message": "Connection to remote server lost",
				}
				if jsonData, err := json.Marshal(errMsg); err == nil {
					localSocket.WriteMessage(websocket.TextMessage, jsonData)
				}
				return
			}

			// Parse and track room state
			var msg map[string]interface{}
			if err := json.Unmarshal(message, &msg); err == nil {
				msgType, _ := msg["type"].(string)
				log.Printf("[PROXY] Remote -> Local (%s): %s", localClientID, msgType)

				state.mu.Lock()
				switch msgType {
				case "ROOM_CREATED":
					if code, ok := msg["code"].(string); ok {
						state.roomID = code
						state.hostID = localClientID
					}
				case "JOIN_SUCCESS":
					if code, ok := msg["code"].(string); ok {
						state.roomID = code
					}
				case "ROOM_CLOSED":
					state.roomID = ""
					state.hostID = ""
				}
				state.mu.Unlock()
			}

			if err := localSocket.WriteMessage(websocket.TextMessage, message); err != nil {
				log.Printf("[LOCAL] Error sending to client %s: %v", localClientID, err)
				return
			}
		}
	}()

	// Handle local -> remote messages
	go func() {
		defer func() {
			remoteSocket.Close()
			localSocket.Close()
			localClients.Delete(localClientID)
		}()

		for {
			_, message, err := localSocket.ReadMessage()
			if err != nil {
				log.Printf("[LOCAL] Client disconnected: %s", localClientID)
				remoteSocket.WriteMessage(websocket.CloseMessage, 
					websocket.FormatCloseMessage(websocket.CloseNormalClosure, "Local client disconnected"))
				return
			}

			// Log message type if JSON
			var msg map[string]interface{}
			if err := json.Unmarshal(message, &msg); err == nil {
				msgType, _ := msg["type"].(string)
				log.Printf("[PROXY] Local -> Remote (%s): %s", localClientID, msgType)
			} else {
				log.Printf("[PROXY] Local -> Remote (%s): [Invalid JSON]", localClientID)
			}

			state.mu.Lock()
			if err := remoteSocket.WriteMessage(websocket.TextMessage, message); err != nil {
				log.Printf("[REMOTE] Error sending from %s: %v", localClientID, err)
			}
			state.mu.Unlock()
		}
	}()
}

func runServer() {
	if config.RemoteURL == "" {
		log.Fatal("Error: REMOTE_URL configuration is required")
	}

	http.HandleFunc("/", handleStatusPage)
	http.HandleFunc("/ws", handleWebSocket)

	addr := ":" + config.LocalPort
	log.Println(strings.Repeat("=", 60))
	log.Println("DevilBridge Local Proxy Server")
	log.Println(strings.Repeat("=", 60))
	log.Printf("Local WebSocket endpoint: ws://localhost%s/ws", addr)
	log.Printf("Local HTTP endpoint: http://localhost%s", addr)
	log.Printf("Remote server: %s", config.RemoteURL)
	log.Println(strings.Repeat("=", 60))
	log.Println("\nProxy is running! Connect your game client to:")
	log.Printf("   ws://localhost%s/ws", addr)
	log.Println("\nStatus page available at:")
	log.Printf("   http://localhost%s", addr)
	log.Println(strings.Repeat("=", 60))

	if err := http.ListenAndServe(addr, nil); err != nil {
		log.Fatal(err)
	}
}

func printUsage() {
	fmt.Printf(`DevilBridge Local Proxy - Configuration Management

Usage: %s [command]

Commands:
  (no args)          Run the proxy server
  help               Show this help message
  config             Show current configuration
  set <key> <value>  Set a configuration value
  get <key>          Get a configuration value
  delete <key>       Delete a configuration value (reset to default)
  reset              Reset all configuration to defaults

Configuration Keys:
  LOCAL_PORT         Local port for the proxy server (default: 8080)
  REMOTE_URL         Remote WebSocket server URL (required)

Examples:
  %s                                    Run the server
  %s config                             Show current config
  %s set REMOTE_URL ws://example.com    Set remote URL
  %s set LOCAL_PORT 9090                Set local port
  %s get REMOTE_URL                     Get remote URL
  %s delete LOCAL_PORT                  Reset local port to default
  %s reset                              Reset all settings

Configuration file location: %s
`, exeName, exeName, exeName, exeName, exeName, exeName, exeName, exeName, configPath)
}

func main() {
	if err := loadConfig(); err != nil {
		log.Printf("Warning: Could not load config: %v", err)
	}

	// Handle CLI commands
	if len(os.Args) > 1 {
		command := strings.ToLower(os.Args[1])
		
		switch command {
		case "help", "-h", "--help":
			printUsage()
			return
			
		case "config":
			fmt.Printf("Configuration file: %s\n\n", configPath)
			fmt.Printf("LOCAL_PORT = %s\n", config.LocalPort)
			fmt.Printf("REMOTE_URL = %s\n", config.RemoteURL)
			return
			
		case "get":
			if len(os.Args) < 3 {
				fmt.Println("Error: Missing key name")
				fmt.Println("Usage: get <key>")
				os.Exit(1)
			}
			key := strings.ToUpper(os.Args[2])
			switch key {
			case "LOCAL_PORT":
				fmt.Println(config.LocalPort)
			case "REMOTE_URL":
				fmt.Println(config.RemoteURL)
			default:
				fmt.Printf("Error: Unknown key '%s'\n", os.Args[2])
				os.Exit(1)
			}
			return
			
		case "set":
			if len(os.Args) < 4 {
				fmt.Println("Error: Missing key or value")
				fmt.Println("Usage: set <key> <value>")
				os.Exit(1)
			}
			key := strings.ToUpper(os.Args[2])
			value := strings.Join(os.Args[3:], " ")
			
			switch key {
			case "LOCAL_PORT":
				config.LocalPort = value
			case "REMOTE_URL":
				config.RemoteURL = value
			default:
				fmt.Printf("Error: Unknown key '%s'\n", os.Args[2])
				os.Exit(1)
			}
			
			if err := saveConfig(); err != nil {
				fmt.Printf("Error saving configuration: %v\n", err)
				os.Exit(1)
			}
			fmt.Printf("Configuration updated: %s = %s\n", key, value)
			return
			
		case "delete":
			if len(os.Args) < 3 {
				fmt.Println("Error: Missing key name")
				fmt.Println("Usage: delete <key>")
				os.Exit(1)
			}
			key := strings.ToUpper(os.Args[2])
			switch key {
			case "LOCAL_PORT":
				config.LocalPort = "8080"
			case "REMOTE_URL":
				config.RemoteURL = ""
			default:
				fmt.Printf("Error: Unknown key '%s'\n", os.Args[2])
				os.Exit(1)
			}
			
			if err := saveConfig(); err != nil {
				fmt.Printf("Error saving configuration: %v\n", err)
				os.Exit(1)
			}
			fmt.Printf("Configuration deleted: %s (reset to default)\n", key)
			return
			
		case "reset":
			config.LocalPort = "8080"
			config.RemoteURL = ""
			if err := saveConfig(); err != nil {
				fmt.Printf("Error saving configuration: %v\n", err)
				os.Exit(1)
			}
			fmt.Println("Configuration reset to defaults")
			return
			
		default:
			fmt.Printf("Error: Unknown command '%s'\n", command)
			fmt.Printf("Run '%s help' for usage information\n", exeName)
			os.Exit(1)
		}
	}

	// Run server (no arguments)
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt)

	go runServer()

	<-sigChan
	log.Println("\nShutting down proxy...")
	
	// Close all connections
	localClients.Range(func(key, value interface{}) bool {
		state := value.(*ClientState)
		if state.remoteSocket != nil {
			state.remoteSocket.Close()
		}
		if state.localSocket != nil {
			state.localSocket.Close()
		}
		return true
	})
	
	log.Println("Proxy stopped.")
}